### ISSUE TITLE
seccomp: block personality(!PER_LINUX)

### ISSUE THREAD

**Issue by topimiettinen** [systemd/systemd#6193]
### Submission type
 - Request for enhancement (RFE)

Docker's default seccomp [profile](https://docs.docker.com/engine/security/seccomp/#significant-syscalls-blocked-by-the-default-profile) blocks `personality(2)` system call with nonzero (not equal to `PER_LINUX`) persona value, because non-Linux emulations may be poorly tested and source of vulnerabilities.

Systemd could do the same. Adding the seccomp filter is trivial, but I can think of a couple of alternative ways to control the filtering:

- max control: introduce a new directive e.g. `SystemCallBlockNonLinuxPersonalities=[true|false]`
- some control: overload `SystemCallArchitectures=native` meaning also that non-native OS personalities are also blocked 
- no control: always block

In my system, I could even block `personality(2)` entirely with SystemCallFilter, except for some user programs like python and pulseaudio. They just call `personality(PER_LINUX)`. So I guess filtering `!PER_LINUX` is relatively safe.

**Comment by poettering** [comment]
I figure RestrictPersonalityDomain=~linux or so would match how we did this for other controls...

**Comment by topimiettinen** [comment]
I used a bit different syntax, is that OK?
RestrictPersonalityDomain=linux

I didn't update man page yet.

**Comment by topimiettinen** [comment]
This version with `LockPersonality` works also with Python and Pulseaudio which failed when `personality(2)` was blocked entirely. The only issue I saw when enabling it for everything was that Chromium couldn't start.

**Comment by tixxdz** [comment]
Hi @topimiettinen,

I think you should allow `personality(0xffffffff)` by default to retrieve the current value, maybe chromium is failing due to that part ?

Also I would suggest you change the D-Bus type of `LockPersonality` or `RestrictPersonalityDomain` from binary to string so it will be easy to extend it later without breaking D-Bus APIs.

@poettering suggested `RestrictPersonalityDomain=~linux` , my preference would be a mix of the two:

`LockPersonality=yes|no|linux|"linux STICKY_TIMEOUTS ..."`

1) yes: locks the Personality= directive as suggested in your man pages. 
2) no: no restrictions
3) linux: Locks the personality to Linux only, this is the same suggestion of @poettering but without the extra "~"
4) "linux STICKY_TIMEOUTS ..." or whatever can be added later if someone needs that. Most users will try to allow flags that do not imply MMAP_PAGE_ZERO, so may be we need a fallback plan to support this.

Right now we only need to support 1, 2 and 3.

Thank you for the patches :-)

**Comment by topimiettinen** [comment]
Chromium just calls `personality(0)`. I suppose it's something else like `NoNewPrivileges` accidentally enabled somewhere.

I made this version which allows 0xffffffff, but it doesn't work:

```
int seccomp_lock_personality(unsigned long personality) {
        _cleanup_(seccomp_releasep) scmp_filter_ctx seccomp = NULL;
        int r;

        seccomp = seccomp_init(SCMP_ACT_ERRNO(EPERM));
        if (!seccomp)
                return -ENOMEM;

        r = seccomp_rule_add_exact(
                        seccomp,
                        SCMP_ACT_ALLOW,
                        SCMP_SYS(personality),
                        1,
                        SCMP_A0(SCMP_CMP_EQ, PERSONALITY_INVALID));
        if (r < 0)
                return r;

        r = seccomp_rule_add_exact(
                        seccomp,
                        SCMP_ACT_ALLOW,
                        SCMP_SYS(personality),
                        1,
                        SCMP_A0(SCMP_CMP_EQ, personality));
        if (r < 0)
                return r;

        return seccomp_load(seccomp);
}

```
What you propose is basically same as my first version (with also "yes" and "no"), but @poettering seemed to prefer simpler version in his comments to 8803799. Maybe you two should sort it out, I have no preference.

The most complex version could be a black/white list of personalities:
`LockPersonality=linux linux_32bit linux_fdpic`
`LockPersonality=~svr4 uw7 # MMAP_PAGE_ZERO`

**Comment by tixxdz** [comment]
@topimiettinen 
> What you propose is basically same as my first version (with also "yes" and "no"), but @poettering seemed to prefer simpler version in his comments to 8803799. Maybe you two should sort it out, I have no preference.

Oh I see, alright then you should go with @poettering suggestion, STICKY_TIMEOUTS is already covered, and for ADDR_NO_RANDOMIZE, READ_IMPLIES_EXEC and READ_IMPLIES_EXEC they should be used in rare cases so I don't think we need to care, or users can just skip `LockPersonality=`, and in this case the D-Bus type can be a boolean.
A D-Bus array is forward compatible, a boolean is a pain, we already made the mistake for some directives, booleans are not extensible and we have to create extra directives... however in this case a boolean is more appropriate for a directive named `LockPersonality=` and as said users/programs can just ignore it.

You should also document that "if running in user mode, or in system mode, but without the CAP_SYS_ADMIN capability (e.g. setting User=), NoNewPrivileges=yes is implied." and also make sure to set no_new_privs here https://github.com/systemd/systemd/blob/master/src/core/execute.c#L1282

Thank you!

**Comment by topimiettinen** [comment]
I couldn't figure out how to make the two rules with libseccomp without turning all existing rules from blacklisting to whitelisting (the default action would be error instead of allow), so I used BPF in 55ec12c.

Pulseaudio and python are fine, the problem with chromium is NNP which prevents the setuid helper from working.

**Comment by poettering** [comment]
Hmm, I'd really prefer if we could use libseccomp for all our rules. Can you explain why libseccomp didn't work for you for this?

Shouldn't it suffice to have a rule that works like this:

```c
seccomp = seccomp_init(SCMP_ALLOW);
seccomp_rule_add_exact(seccomp, SCMP_ACT_ALLOW,
                        SCMP_SYS(personality),
                        1,
                        SCMP_A0(SCMP_CMP_EQ, PERSONALITY_INVALID));
seccomp_rule_add_exact(seccomp, SCM_ACT_ALLOW,
                        SCMP_SYS(personality),
                        1,
                        SCMP_A0(SCMP_CMP_EQ, personality);
seccomp_rule_add_exact(seccomp, SCMP_ACT_ERRNO(EPERM), SCMP_SYS(personality), 0);
```

i.e. permit all syscalls by default, and for personality() refuse all invocations, except the right personality and the one value for querying it?

**Comment by topimiettinen** [comment]
I don't like the BPF version much either, it's just too crude even if there's full control of the execution.

I didn't find a good description of the seccomp filter logic and libseccomp code or docs don't help either. I think that all the filter rules are evaluated and only then the result is checked, highest priority wins. A different logic could be that when there is a match, the rest of the rules are ignored, but this does not seem to be the case.

But I tested your libseccomp version, it does not work:

```
+int seccomp_lock_personality(unsigned long personality) {
+        _cleanup_(seccomp_releasep) scmp_filter_ctx seccomp = NULL;
+        int r;
+
+        seccomp = seccomp_init(SCMP_ACT_ALLOW);
+        if (!seccomp)
+                return -ENOMEM;
+
+        r = seccomp_rule_add_exact(seccomp, SCMP_ACT_ALLOW,
+                                   SCMP_SYS(personality),
+                                   1,
+                                   SCMP_A0(SCMP_CMP_EQ, PERSONALITY_INVALID));
+        if (r < 0)
+                return r;
+
+        r = seccomp_rule_add_exact(seccomp, SCMP_ACT_ALLOW,
+                               SCMP_SYS(personality),
+                               1,
+                               SCMP_A0(SCMP_CMP_EQ, personality));
+        if (r < 0)
+                return r;
+
+        r = seccomp_rule_add_exact(seccomp, SCMP_ACT_ERRNO(EPERM),
+                                   SCMP_SYS(personality), 0);
+        if (r < 0)
+                return r;
+
+        r = seccomp_attr_set(seccomp, SCMP_FLTATR_CTL_NNP, 0);
+        if (r < 0)
+                return r;
+
+        return seccomp_load(seccomp);
+}

```
I suppose the problem is that the default rule of `SCMP_ACT_ALLOW` means that only the error actions count as the allow actions don't have any effect, so this effectively just blocks all personality calls.

When I changed the default rule to `SCMP_ACT_ERRNO(EPERM)`, no system call passed through after installing this filter. So I suppose for this way to work, all other filters in systemd should be rewritten as allow rules, ugh.

Does this make any sense to you? I'm making too many guesses and assumptions.

**Comment by poettering** [comment]
Hmm, maybe @pcmoore has a suggestion? @pcmoore, any chance you can have a look at the last code fragment @topimiettinen pasted, and maybe let us know why this doesn't work properly? the idea is that this prohibits all invocations of the personality() syscall, except if the parameter passed is 0xffffffff or the personality actually configured.

**Comment by pcmoore** [comment]
Hello all,

Using the example code that @topimiettinen posted above, the problem is that you are creating a seccomp filter with a default rule of `SCMP_ACT_ALLOW`:

    +        seccomp = seccomp_init(SCMP_ACT_ALLOW);
    +        if (!seccomp)
    +                return -ENOMEM;

... and then trying to create a filter rule with the same action, `SCMP_ACT_ALLOW`:

    +        r = seccomp_rule_add_exact(seccomp, SCMP_ACT_ALLOW,
    +                                   SCMP_SYS(personality),
    +                                   1,
    +                                   SCMP_A0(SCMP_CMP_EQ, PERSONALITY_INVALID));
    +        if (r < 0)
    +                return r;

... from a seccomp/libseccomp perspective this is a pointless/no-op rule so it returns an error.  When you add a rule to a filter using libseccomp the rule's action must be different from the default action of the filter.

Hopefully this helps, if you have any more questions just ping me.

**Comment by topimiettinen** [comment]
This confirms what I thought. I think it's not possible to add both personality checks with libseccomp unless we rewrite all other filters to whitelisting logic. That could actually be an improvement.

**Comment by poettering** [comment]
@topimiettinen what do you mean by "all other filters"? note that we use separate seccomp objects for all filters we implement, and unless I am mistaking they should all be independent of each other then...

I figure the following should work: install a filter with a default of ALLOW, and two rule:

```c
        r = seccomp_rule_add_exact(seccomp, SCMP_ACT_ERRNO(EPERM),
                                   SCMP_SYS(personality),
                                   2,
                                   SCMP_A0(SCMP_CMP_GT, personality),
                                   SCMP_A0(SCMP_CMP_LT, PERSONALITY_INVALID));
        r = seccomp_rule_add_exact(seccomp, SCMP_ACT_ERRNO(EPERM),
                                   SCMP_SYS(personality),
                                   1,
                                   SCMP_A0(SCMP_CMP_LT, personality));
```
i.e. mask out the range `< personality`, and the range `> personality` (but only up to PERSONALITY_INVALID), and leave everything else untouched... (of course, this should be conditionalized a bit for the case where personality == 0)...

**Comment by topimiettinen** [comment]
@poettering: For example `SystemCallFilter` filters would be created so that the default rule would be error and then we add the allowed system calls. It should be doable with set operations. Though I suppose if there are further rules affecting a system call (like `RestrictAddressFamilies`), the logic will be more complex.

I don't think the objects are independent, setting default rule to error in one object seems to affect the rest.

Your version doesn't work, I guess you can't make rules which check the same register. It does work if I take out that part (ifdeffed out):

```
+int seccomp_lock_personality(unsigned long personality) {
+        _cleanup_(seccomp_releasep) scmp_filter_ctx seccomp = NULL;
+        int r;
+
+        seccomp = seccomp_init(SCMP_ACT_ALLOW);
+        if (!seccomp)
+                return -ENOMEM;
+
+        if (personality > 0) {
+                r = seccomp_rule_add_exact(seccomp, SCMP_ACT_ERRNO(EPERM),
+                                           SCMP_SYS(personality),
+                                           1,
+                                           SCMP_A0(SCMP_CMP_LT, personality));
+                if (r < 0)
+                        return r;
+        }
+
+#if 0
+        r = seccomp_rule_add_exact(seccomp, SCMP_ACT_ERRNO(EPERM),
+                                   SCMP_SYS(personality),
+                                   2,
+                                   SCMP_A0(SCMP_CMP_GT, personality),
+                                   SCMP_A0(SCMP_CMP_LT, PERSONALITY_INVALID));
+        if (r < 0)
+                return r;
+#endif
+
+        r = seccomp_rule_add_exact(seccomp, SCMP_ACT_ERRNO(EPERM),
+                                   SCMP_SYS(personality),
+                                   1,
+                                   SCMP_A0(SCMP_CMP_GT, PERSONALITY_INVALID));
+        if (r < 0)
+                return r;
+
+        return seccomp_load(seccomp);
+}
```

 @pcmoore?

**Comment by topimiettinen** [comment]
It's easiest for now to skip the check for `PERSONALITY_INVALID`. I didn't see any (non-NNP) problems with this version (65ef2a9) which only allows current personality, implemented with libseccomp.

If one day libseccomp gains more complex logic or we decide to convert all seccomp rules to whitelisting (or BPF), this can be improved.

Also now the default personality (Linux) is used for locking, if `Personality` directive is not used but LockPersonality is enabled.

**Comment by tixxdz** [comment]
@topimiettinen https://github.com/systemd/systemd/commit/65ef2a93883c8709c2cab86107560e22afe0fb34 lgtm, hmm maybe later a test to assert that "LockPersonality=" works as intended.

Thank you!

**Comment by poettering** [comment]
this really should have a test right from the beginning. We had problems with the other seccomp bits that they didn't actually work as intended after mergin them without tests. seccomp isn't trivial to make use of, hence errors are easily made, and to counter them this should have tests from the beginning really.

**Comment by topimiettinen** [comment]
I added some tests with 3ed277f, also the comment suggested by @tixxdz and corrected the output to match the changed logic.

**Comment by poettering** [comment]
@topimiettinen btw, can you maybe submit your stuff as PR to systemd? it's much easier to review then, as we can make proper use of the review logic github provides. It's fine if the code isn't fully ready yet when you do, after all it's kind the point of the exercise to make it perfect while the PR remains unmerged

**Comment by topimiettinen** [comment]
There is #6256. I updated the title and description so it matches the commit.

**Comment by topimiettinen** [comment]
Should I make a new PR or is there something else you want to change?

**Comment by topimiettinen** [comment]
Closed by 40cdf0c.
